core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NoticeContainer.java

```diff
@@ -44,18 +45,31 @@
    *
    * <p>Note that system errors are not limited since we don't expect to have a lot of them.
    */
-  private static final int MAX_VALIDATION_NOTICES = 10000000;
+  private static final int MAX_VALIDATION_NOTICES = 10_000_000;
 
   private final List<ValidationNotice> validationNotices = new ArrayList<>();
   private final List<SystemError> systemErrors = new ArrayList<>();
   private boolean hasValidationErrors = false;
 
   /** Adds a new validation notice to the container (if there is capacity). */
   public void addValidationNotice(ValidationNotice notice) {
+    addValidationNotice(notice, MAX_VALIDATION_NOTICES, MAX_PER_NOTICE_TYPE_AND_SEVERITY);
+  }
+
+  /** Adds a new validation notice to the container (if there is capacity). */
+  public void addValidationNotice(
+      ValidationNotice notice, int maxTotalValidationNotices, int maxValidationNoticePerType) {
     if (notice.isError()) {
       hasValidationErrors = true;
     }
-    if (validationNotices.size() <= MAX_VALIDATION_NOTICES) {
+    if (validationNotices.size() >= maxTotalValidationNotices) {
+      return;
+    }
+    ListMultimap<String, ValidationNotice> noticesOfType =
+        groupNoticesByTypeAndSeverity(validationNotices);
```

Comment:
A couple of things

1. This method runs O(N) which means that adding N notices will work O(N^2). With 10M limit this will take a good portion of time to run. 
2. I know that we're running validation rules in parallel. Does NoticeContainer need to be thread safe? 

To solve the first one, let's just count number of notices with same type and severity in a Map<String, Integer>. And then add the notice to validationNotices if both of counters (total and per severity+type) are fine. 

As for the second we can just add synchronized to the method signature.